modules/utils/graph/Redif.cpp

```diff
@@ -0,0 +1,325 @@
+			}
+			/// Check if Model is able to denoise Test Posize_t
+			blaze::DynamicVector<Tv, columnVector> temp = column(distTotal, l_idx[i]);
+			for (size_t k = 0; k < idxVec.size(); k++) idxVec[k] = k;
+			Quicksort(idxVec, temp, 0, idxVec.size() - 1);
+			if (oneDist > temp[k_check - 1])
+				cout << "Error. Testposize_t of Index " << i << " cannot be denoised based on underlying model.\n";
+		}
+
+		blaze::DynamicMatrix<Tv> xEncoded = xTrain;
+		/// Iterative backward Diffusion process
+		for (size_t i = 0; i < nX; i++)
+			for (size_t j = 0; j < xEncoded.columns(); j++)
+				xEncoded(l_idx[i], j) = x(i, j);
+
+		for (size_t i = 0; i < nIter; i++)
+		{
+			blaze::DynamicMatrix<Tv> D(nTrain, nTrain);
+			for (size_t k = 0; k < D.rows(); k++)
+				D(k, k) = LArray[i](k, k);
+			xEncoded = blaze::inv(D + 0.25*LArray[i])*(D*xEncoded);
+		}
+
+		return xEncoded;
+	}
+
+	template <typename recType, class Metric, typename Tv>
+	blaze::DynamicMatrix<Tv> Redif<recType, Metric, Tv>::decode(blaze::DynamicMatrix<Tv> xEncoded, blaze::DynamicMatrix<Tv> xTrainEncoded, blaze::DynamicVector<size_t> l_idx)
+	{
+		size_t n = xTrainEncoded.rows();
+		size_t nIter = LArray.size();
+		for (size_t i = 0; i < l_idx.size(); i++)
+			for (size_t j = 0; j < xTrainEncoded.columns(); j++)
+				xTrainEncoded(l_idx[i], j) = xEncoded(i, j);
+
+		for (size_t i = 0; i < nIter; i++)
+		{
+			blaze::DynamicMatrix<Tv> temp = LArray[nIter - i - 1];
+			cout << temp.rows() << " X " << temp.columns() << endl;
+			blaze::DynamicMatrix<Tv> D(n, n);
+			for (size_t k = 0; k < n; k++)
+				D(k, k) = temp(k, k);
+			xTrainEncoded = blaze::inv(D)*(D + 0.25*LArray[nIter - i - 1])*xTrainEncoded;
+		}
+		blaze::DynamicMatrix<Tv> xDecoded = blaze::DynamicMatrix<Tv>(l_idx.size(), xTrainEncoded.columns());
+		for (size_t i = 0; i < xDecoded.rows(); i++)
+			for (size_t j = 0; j < xDecoded.columns(); j++)
+				xDecoded(i, j) = xTrainEncoded(l_idx[i], j);
+
+		return xDecoded;
+	}
+
+	template <typename recType, class Metric, typename Tv>
+	template<bool flag>
+	size_t Redif<recType, Metric, Tv>::partition(blaze::DynamicVector<size_t> & idx, blaze::DynamicVector<Tv, flag> & val, size_t start, size_t end)
+	{
+		double pivot = val[end];
+
+		size_t P_index = start;
+		size_t i;
+		size_t idxTemp;
+		double valTemp;
+
+		for (i = start; i < end; i++)
+		{
+			if (val[i] <= pivot)
+			{
+				valTemp = val[i];
+				val[i] = val[P_index];
+				val[P_index] = valTemp;
+
+				idxTemp = idx[i];
+				idx[i] = idx[P_index];
+				idx[P_index] = idxTemp;
+
+				P_index++;
+			}
+		}
+
+		valTemp = val[end];
+		val[end] = val[P_index];
+		val[P_index] = valTemp;
+
+		idxTemp = idx[end];
+		idx[end] = idx[P_index];
+		idx[P_index] = idxTemp;
+
+		return P_index;
+	}
+
+	template <typename recType, class Metric, typename Tv>
+	template<bool flag>
+	void Redif<recType, Metric, Tv>::Quicksort(blaze::DynamicVector<size_t> & idx, blaze::DynamicVector<Tv, flag> & data, size_t start, size_t end)

```

Comment:
why own quicksort? it is incredible slow, when implementing on its own in comparison to STL version, and it blows the code size.